usbd: Add basic HID keypad (with status LED detection), basic mass storage

turmoni opened the pull request:
Creating as a draft pull request because I don't think this is really ready to be merged as-is, since I was mostly working on just getting it working for me and I imagine the structure could do with being improved, especially when it comes to where to put constants and the like. Points of note:

- I think there has been some confusion between IN and OUT - see my modifications to `hid.py` where I've swapped them around.
- The mass storage is very limited, and only implements what I wanted, which is basically read-only, a single LUN, and with the entire filesystem as a variable in RAM. (Update: now also supports objects using the AbstractBlockDev protocol.)
- Some of this code can eventually be removed since it's there to work around issues I failed to figure out (see `_retry_xfer_cb` in `device.py` and `try_to_prepare_cbw` in `msc.py` for examples)

I also know that I'm not following micropython-lib's conventions for submitting a PR, but since I'm submitting this to you and I'm not necessarily expecting it to be directly merged I hope that's less important. At least, I made my commit messages and pushed them up before reading that.

projectgus committed:
usbd: Add USB device drivers implemented in Python.

Rely on support implemented in the machine.USBD() object on the MicroPython
side, providing a thin wrapper around TinyUSB "application" device class
driver.

projectgus committed:
usbd: Add USB device drivers implemented in Python.

Rely on support implemented in the machine.USBD() object on the MicroPython
side, providing a thin wrapper around TinyUSB "application" device class
driver.

projectgus committed:
usbd: Add midi interface definition from @paulhamsh.

Based on https://github.com/paulhamsh/Micropython-Midi-Device
as of commit 2678d13.

With additions/edits by me.

projectgus committed:
usbd: Add midi interface definition from @paulhamsh.

Taken from https://github.com/paulhamsh/Micropython-Midi-Device
as of commit 2678d13.

projectgus committed:
usbd: Major cleanup, refactor.

- Add micropython-lib 'usbd' package (provisional).
- Update midi implementation a bit.
- Rearrange code to work with package structure
- Convert docstrings to regular comments to save flash.

turmoni committed:
Add basic keypad support

turmoni committed:
Fix report count, remove irrelevant comments

turmoni committed:
Add basic, read-only MSC support, and add LED status to keypad.

turmoni committed:
Actually add the changes methoned in the previous commit message, and add more documentation to msc.py

andrewleech commented:
Impressive effort getting this going, looks great! With the mass storage, have you had it running against a "real" flash filesystem on the device? Does it need / work with a Block device eg one that's formatted as FAT?

turmoni commented:
> Impressive effort getting this going, looks great! With the mass storage, have you had it running against a "real" flash filesystem on the device? Does it need / work with a Block device eg one that's formatted as FAT?

Thanks! I've realised that "filesystem" is really a bit of a misnomer, since it doesn't know or care about the filesystem, it just needs something it can operate on. Currently it needs something that works like a bytestring, but I imagine it wouldn't be too hard to extend it to do a type check on whatever it is you give it and access it appropriately. I would say it would probably be possible to create an object to pass in to it that would work transparently, but micropython doesn't support the `__bytes__` method on classes so I'm not sure how you'd do that. All the changes that would be needed to be made would be in the `StorageDevice` class though, since the upper layers don't actually care about anything but translating between the host and the (virtual) SCSI device.

turmoni commented:
You've successfully got me into "just quickly adding" the ability to use `rp2.Flash`, which exposed a bug, and my fix has exposed other bugs, and so I guess I'll be figuring out how to fix those. I have got the main functionality working though, I can `cat /dev/sde` and (I think?) get the littlefs filesystem on my Pico.

(This chain of bugs brought to you by Linux asking for 30 4096-byte blocks in one command, which I hadn't seen before due to my filesystem entirely fitting in RAM.)

andrewleech commented:
Nice one! Yeah my biggest issue with MSC is that it just exposes the underlying block device and filesystem - which for me is always littlefs so not easily supported on computer.

I'm actually keen to use this new USB support to add MTP as used on cameras and Android phones - this operates at the file level so should be good on any filesystem.

turmoni committed:
usbd: Bugfixes around data transfer, support using an AbstractBlockDev-based device

turmoni commented:
I've now added support for AbstractBlockDev-based devices, plus the required bug fixes. I'm not sure it's done in the neatest way, but it does seem to reliably work

projectgus commented:
Hi @turmoni,

This looks really great, thanks!

> I think there has been some confusion between IN and OUT - see my modifications to hid.py where I've swapped them around.

Good catch on this, I'm not sure how I got this far without noticing I had them backwards!

I'm keen to bring quite a bit of this into the main micropython-lib PR. When you get a minute, do you mind going back to pop some Copyright & License notices on the new files? Feel free to add yourself to the ones you fixed/modified, as well.

turmoni committed:
Merge remote-tracking branch 'upstream/feature/usbd_python' into feature/usbd_python

turmoni committed:
usbd: Add copyright notices (+delete file that has gone from upstream but wasn't removed in the merge)

turmoni committed:
usbd: Run "black" with the right options for the style checker to be happy

turmoni committed:
usbd: Use EP_IN_FLAG from utils for mass storage

turmoni committed:
usbd: Re-run black to fix the missing comma

turmoni commented:
Thanks! I've added those to the files I created now (and got rid of my own EP constants and some other bits to adjust to the changes you've made), I'm happy to make any changes here (or in a new branch) to make it easier if needed.

projectgus commented:
@turmoni Thanks dave, that should be heaps for now. I still need to figure out exactly how best to structure the PRs into micropython-lib (to not add too much in one PR, and to correctly group things like examples or modules for particular functionality). So I'm happy to pick and choose from your branch as needed at the moment.

If you did have time, I'd be very curious if your MSC class still needs the extra workaround of `_always_cb` or if it now works OK on Windows without it. If you don't have time then I'll get around to testing this myself soon as well, though.

turmoni commented:
I cleaned that bit up on seeing the changes you'd made and it seems like it works fine without it now - my fork doesn't have it any more. At least, I currently have it connected up with the `rp2.Flash` object and I can see many debug logs continuing to flood away in the background, even if Windows tells me the filesystem needs to be formatted before it can be used, and on Linux I can happily `cat` the device and get everything off it.

projectgus committed:
usbd: Add support for configuration open and reset callbacks.

Implement by overriding USBInterface.handle_open or handle_reset.

projectgus committed:
usbd: Add USB interface functions for endpoint STALL support.

Necessary for MSC device class, possibly other purposes.

projectgus committed:
usbd: Implement SET_REPORT support for OUT direction HID data.

projectgus committed:
usbd: Rename ustruct->struct.

projectgus committed:
usbd: Add hid keypad example from @turmoni .

As contributed in https://github.com/projectgus/micropython-lib/pull/1 commit 5c51a9ee70

This version of the hidkeypad module depends on some other code changes from the
linked PR that aren't included here, so it won't work here yet.

projectgus committed:
usbd: Update hid_keypad example module.

- Use SET_REPORT via control transfer.
- Merge the keycodes module into hid_keypad.
- Remove LEDs other than NumLock to make the report handler even simpler.

projectgus committed:
usbd: Implement ruff, black linter & formatting fixes.

projectgus committed:
usbd: Add missing manifest file.

turmoni committed:
Merge remote-tracking branch 'upstream/feature/usbd_python' into feature/usbd_python

turmoni committed:
usbd: Theoretically handle resets and bad CBWs better in msc

turmoni commented:
Right, I _think_ I have added decent usage of the new stall options - unfortunately I think my code ended up getting robust enough and my memory of how to do this stuff has degraded enough that I'm not _entirely_ sure how to test it properly, so I'm by no means guaranteeing that this will appropriately behave in response to a reset or an invalid/non-meaningful CBW.
